app/src/main/java/org/thoughtcrime/securesms/payments/PaymentParcelable.java

```
@@ -0,0 +1,215 @@
+    private final Direction       direction;
+    private final State           state;
+    private final FailureReason   failureReason;
+    private final String          note;
+    private final Money           amount;
+    private final Money           fee;
+    private final PaymentMetaData paymentMetaData;
+    private final boolean         isSeen;
+    private final Money           amountWithDirection;
+    private final Money           amountPlusFeeWithDirection;
+    private final boolean         isDefrag;
+
+    private ParcelPayment(Parcel in) {
+      try {
+        uuid = UUID.fromString(in.readString());
+
+        PayeeParcelable payeeParcelable = in.readParcelable(PayeeParcelable.class.getClassLoader());
+        payee = payeeParcelable.getPayee();
+
+        blockIndex                 = in.readLong();
+        blockTimestamp             = in.readLong();
+        timestamp                  = in.readLong();
+        displayTimestamp           = in.readLong();
+        direction                  = Direction.deserialize(in.readInt());
+        state                      = State.deserialize(in.readInt());
+
+        int failureReasonSerialized = in.readInt();
+        if (failureReasonSerialized == -1) {
+          failureReason = null;
+        } else {
+          failureReason = FailureReason.deserialize(failureReasonSerialized);
+        }
+
+        note                       = in.readString();
+        amount                     = Money.parse(in.readString());
+        fee                        = Money.parse(in.readString());
+        paymentMetaData            = PaymentMetaData.parseFrom(in.createByteArray());
+        isSeen                     = in.readByte() == 1;
+        amountWithDirection        = Money.parse(in.readString());
+        amountPlusFeeWithDirection = Money.parse(in.readString());
+        isDefrag                   = in.readByte() == 1;
+      } catch (Money.ParseException | InvalidProtocolBufferException e) {
```

Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)